crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,1179 @@
+                let Some(module_name) = Self::absolute_path_to_module_name(path, search_path) else {
+                    return false;
+                };
+                let typeshed_versions = Self::load_typeshed_versions(db, stdlib_root);
+                match typeshed_versions.query_module(&module_name, get_target_py_version(db)) {
+                    TypeshedVersionsQueryResult::Exists
+                    | TypeshedVersionsQueryResult::MaybeExists => {
+                        db.file_system().exists(&path.join("__init__.pyi"))
+                    }
+                    TypeshedVersionsQueryResult::DoesNotExist => false,
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    fn to_vfs_file(self, db: &dyn Db, search_path: ModuleResolutionPathRef<'a>) -> Option<VfsFile> {
+        match (self, search_path.0) {
+            (Self::Extra(path), Self::Extra(_)) => system_path_to_file(db.upcast(), path),
+            (Self::FirstParty(path), Self::FirstParty(_)) => system_path_to_file(db.upcast(), path),
+            (Self::SitePackages(path), Self::SitePackages(_)) => {
+                system_path_to_file(db.upcast(), path)
+            }
+            (Self::StandardLibrary(path), Self::StandardLibrary(stdlib_root)) => {
+                let module_name = Self::absolute_path_to_module_name(path, search_path)?;
+                let versions = Self::load_typeshed_versions(db, stdlib_root);
+                match versions.query_module(&module_name, get_target_py_version(db)) {
+                    TypeshedVersionsQueryResult::Exists
+                    | TypeshedVersionsQueryResult::MaybeExists => {
+                        system_path_to_file(db.upcast(), path)
+                    }
+                    TypeshedVersionsQueryResult::DoesNotExist => None,
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn to_module_name(self) -> Option<ModuleName> {
+        let (fs_path, skip_final_part) = match self {
+            Self::Extra(path) | Self::FirstParty(path) | Self::SitePackages(path) => (
+                path,
+                path.ends_with("__init__.py") || path.ends_with("__init__.pyi"),
+            ),
+            Self::StandardLibrary(path) => (path, path.ends_with("__init__.pyi")),
+        };
+
+        let parent_components = fs_path
+            .parent()?
+            .components()
+            .map(|component| component.as_str());
+
+        if skip_final_part {
+            ModuleName::from_components(parent_components)
+        } else {
+            ModuleName::from_components(parent_components.chain(fs_path.file_stem()))
```

Comment:
Nice, that's even better than a custom iterator!

Reply:
yeah, I realised yesterday that I was overcomplicating it 😄 your suggestion of adding a `ModuleName::from_components()` method made it much easier to see how to do it more simply :)